.github/bin/run-clang-tidy.sh

```diff
@@ -76,96 +110,109 @@ fi
+for f in $(cat ${FILES_OF_INTEREST}); do
+  if [ "${f##*.}" == "h" ]; then     # only headers expected to be relevant
+    c_hash="$(cat $f | md5sum | awk '{print $1}')"
+    echo "s|$f|$c_hash|g" | sed 's|\.|\\.|g' >> "${INCLUDE_REPLACE}"
+  fi
+done
 
-EXIT_CODE=0
-if [ -s ${FILES_TO_PROCESS} ]; then
-  echo "::group::Run ${PARALLEL_COUNT} parallel invocations of ${CLANG_TIDY} in chunks of ${FILES_PER_INVOCATION} files."
+# Create a content hash of each file, first changing include filenames with
+# their content-hash as described above.
+# Make a list of all files we don't have hashes for yet.
+for f in $(cat ${FILES_OF_INTEREST}); do
+  c_hash="$(sed -f "${INCLUDE_REPLACE}" < $f | md5sum | awk '{print $1}')"
+  echo "${c_hash}" >> "${HASHES_OF_INTEREST}"
+  if [ ! -r "${CONTENT_DIR}/${c_hash}" ]; then  # don't have, need to create
+    echo "${c_hash} ${f}" >> "${FILE_HASHES_TO_PROCESS}"
+  else
+    touch "${CONTENT_DIR}/${c_hash}"  # keep current to prevent garbage collect
+  fi
+done
 
-  cat ${FILES_TO_PROCESS} \
-    | xargs -P${PARALLEL_COUNT} -n ${FILES_PER_INVOCATION} -- \
-      ${CLANG_TIDY} --config="$(cat ${CLANG_CONFIG_FILE})" --quiet "${ADDITIONAL_TIDY_OPTIONS[@]}" 2>/dev/null \
-    | sed -e "s@${CANONICALIZE_SOURCE_PATH_REGEX}@@g" \
-    > ${TIDY_OUT}.tmp
+echo "$(wc -l < ${HASHES_OF_INTEREST}) files of interest."
 
-  mv ${TIDY_OUT}.tmp ${TIDY_OUT}
-  cat ${TIDY_OUT}
+if [ -s "${FILE_HASHES_TO_PROCESS}" ]; then
+  echo "::group::$(wc -l < ${FILE_HASHES_TO_PROCESS}) files to process"
+  cat "${FILE_HASHES_TO_PROCESS}"
   echo "::endgroup::"
 
-  echo ::group::Summary
-  sed -e 's|\(.*\)\(\[[a-zA-Z.-]*\]$\)|\2|p;d' < ${TIDY_OUT} | sort | uniq -c | sort -rn
+  # A compilation database is needed to run clang tidy.
+  echo ::group::Build compilation database
+  $(dirname $0)/make-compilation-db.sh
   echo "::endgroup::"
-else
-  echo "Skipping clang-tidy run: nothing to do"
+
+  # The files might be reported with various absolute prefixes that
+  # we remove to canonicalize and report as relative paths from project
+  # root.
+  readonly EXEC_ROOT="$(bazel info execution_root)"
+  readonly CURRENT_DIR="$(pwd)"
+  readonly CANONICALIZE_SOURCE_PATH_REGEX="^\(${EXEC_ROOT}\|${CURRENT_DIR}\|\.\)/"
+
+  # Call an in-line shell script in parallel that gets the file/hash tuple
+  # as parameter. Split the argument into ${args[0]} (hash value)
+  # and ${args[1]} (cc-file to process) and call clang-tidy on it. So each
+  # clang-tidy result is stored to its content-addressed file.
+  # (shell-quoting nightmare ahead:)
+  echo "Processing in ${PARALLEL_COUNT} parallel invocations of ${CLANG_TIDY}"
+  cat "${FILE_HASHES_TO_PROCESS}" | \
+    xargs -P"${PARALLEL_COUNT}" -d'\n' -n1 -- \
+          bash -c "read -a args <<< \$1 ;
+              ${CLANG_TIDY} --config=\"$(cat ${CLANG_CONFIG_FILE})\" \
+              --quiet ${ADDITIONAL_TIDY_OPTIONS} \${args[1]} 2>/dev/null \
```

Comment:
For arrays, normally this needs to be expanded as `"${ADDITIONAL_TIDY_OPTIONS[@]}"`.

Since you are inside a here-doc, you want just `${ADDITIONAL_TIDY_OPTIONS[@]}` or with additional quote protection, `${ADDITIONAL_TIDY_OPTIONS[@]@Q}`.  (I just learned this latter variant recently.)

Reply:
Done. The quoting-hell prevented that to work inside the shell script, so generated a helper shell-script in a here document that then is executed.